### PR TITLE
Implement live reconfiguration in the compute_ctl, rebased

### PR DESCRIPTION
**Pull request by hlinnaka**
This is PR #3923, rebased over #3938.

I tried to refrain from making any non-trivial changes, but here are a few notable ones:

- ComputeState was quite confusing, as it was used both to define the format of the JSON that the `/get_status` API returned, and to hold other mutable state of a ComputeNode. I refactored it so that ComputeState defines the format of `/get_status`, like on 'main', and there's another struct, `ComputeNodeInner`, which holds all the mutable state.
- I created yet another struct, `ParsedSpec`, which encapsulates `ComputeSpec` and the fields derived from it.

### CODE REVIEW

File: compute_tools/src/compute.rs [hunk 1]
```diff
@@ -42,25 +42,92 @@ pub struct ComputeNode {
...
+    /// Coupled with `Condvar` to allow notifying HTTP API and configurator
+    /// thread about state changes. To allow HTTP API server to serving status
+    /// requests, while configuration is in progress, lock should be held only
+    /// for short periods of time to do read/write, not the whole configuration
+    /// process.
+    pub inner: Mutex<ComputeNodeInner>,
+    pub state_changed: Condvar,
+}
+
+pub struct ComputeNodeInner {
```
Comment (ololobus) [inline compute_tools/src/compute.rs:64]: Separate `state_changed` looks good to me. I'm not sure about additional `ComputeNodeInner`, it's still a volatile compute state, but this

> ComputeState was quite confusing, as it was used both to define the format of the JSON that the /get_status API returned, and to hold other mutable state of a ComputeNode.

indeed a good point. Yet, I'd maybe solve it in a bit different way -- define `GET /status` response as a separate structure, which can be built from `ComputeState`. It should eliminate confusion and will be generally cleaner than currently in the main. I've already defined struct for error response

About metrics, I'm 100% +1 for moving them under the same lock, but as you properly pointed out in the original PR, let's keep subset of changes smaller to simplify review

If no objections, I'd incorporate mentioned above changes into https://github.com/neondatabase/neon/pull/3963

File: compute_tools/src/http/api.rs [hunk 2]
```diff
@@ -12,6 +15,44 @@ use serde_json;
 use tracing::{error, info};
 use tracing_utils::http::OtelName;
 
+async fn handle_spec_request(req: Request<Body>, compute: &Arc<ComputeNode>) -> Result<(), (String, StatusCode)> {
```
Comment (ololobus) [inline compute_tools/src/http/api.rs:18]: Refactored this in a similar way too, thanks. Also tweaked waiting part to spawn a separate blocking thread with `tokio::task::spawn_blocking` to do not block main pool of worker threads and serve status requests. I didn't expect it to be a problem, but it seems to be a safer approach

File: compute_tools/src/compute.rs (hunk 1 above)
Comment (hlinnaka) [reply compute_tools/src/compute.rs:64]: >  Yet, I'd maybe solve it in a bit different way -- define GET /status response as a separate structure, which can be built from ComputeState.

Works for me

**Comment by ololobus** [comment]
Separated the last part of the original PR into https://github.com/neondatabase/neon/pull/3980, so I think we can close this one

**Pull request closed without merging**